### PR TITLE
Support cache control headers

### PR DESCRIPTION
**Pull request by seladb**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

### CODE REVIEW

File: starlite/handlers/http.py [hunk 1]
```diff
@@ -424,10 +424,10 @@ def resolve_response_headers(self) -> "ResponseHeadersMap":
         resolved_response_headers = {}
         for layer in self.ownership_layers:
             resolved_response_headers.update(layer.response_headers or {})
-        if self.cache_control:
-            resolved_response_headers.update(
-                {CacheControlHeader.CACHE_CONTROL_HEADER: ResponseHeader(value=self.cache_control.to_header())}
-            )
+            if cache_control := getattr(layer, "cache_control", None):
```
Comment (provinzkraut) [inline starlite/handlers/http.py:427]: Probably should be 

```python
cache_control = getattr(layer, "cache_control", None)
if cache_control:
   ...
```

To support Python < 3.8

Comment (seladb) [reply starlite/handlers/http.py]: Thanks, fixed!

File: starlite/datastructures/headers.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+from starlite.exceptions import ImproperlyConfiguredException
+
+if TYPE_CHECKING:
+    from typing import Any, Dict
+
+
+class Header(BaseModel, ABC):
+    """An abstract type for HTTP headers."""
+
```
Comment (provinzkraut) [inline starlite/datastructures/headers.py:14]: The `documentation_only` field is missing. You mention it in the documentation, so it should be added here as well.

File: tests/response/test_response_headers.py [hunk 3]
```diff
@@ -72,3 +73,27 @@ def my_handler(data: Dict[str, str]) -> Dict[str, str]:
         response = client.post("/test", json={"hello": "world"})
         assert response.status_code == HTTP_201_CREATED
         assert response.headers.get("test-header") == "test value"
+
+
```
Comment (provinzkraut) [inline tests/response/test_response_headers.py:77]: This should probably include tests for overwrite behaviour between `response_headers=` and `cache_control=`. It's mentioned in the docs, so accidentally changing the overwrite cascade in some way could break things if people rely on the specific order.

File: tests/datastructures/test_headers.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+
+def test_cache_control_to_header() -> None:
+    header = CacheControlHeader(max_age=10, private=True)
+    expected_header_values = ["max-age=10, private", "private, max-age=10"]
+    assert header.to_header() in expected_header_values
+    assert header.to_header(include_header_name=True) in [f"cache-control: {v}" for v in expected_header_values]
+
+
+def test_cache_control_from_header() -> None:
```
Comment (provinzkraut) [inline tests/datastructures/test_headers.py:14]: Maybe test all possible fields here to make sure they're included / of the right type after parsing

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Looks mostly good, I just left a few minor notes regarding documentation and tests.

The `documentation_only` field is missing though, which should definitely be added.

**Review by Goldziher — APPROVED** [review]
Great stuff. I have no comments on top of what @provinzkraut already wrote. Merge when ready.

File: docs/usage/5-responses/4-response-headers.md [hunk 5]
```diff
@@ -228,3 +228,56 @@ router = Router(
...
+
+@get("/population")
+def get_population_count() -> int:
+    return 100000
+
+
+router = Router(
+    route_handlers=[MyController, get_population_count],
+    cache_control=CacheControlHeader(max_age=2_628_288, public=True),
+)
```
Comment (peterschutt) [inline docs/usage/5-responses/4-response-headers.md:273]: As this looks like a complete example, please have a look at #622 for how we prefer to include complete examples in the docs.

You can write the example as a script, include a test for it, and then cross reference the script in the docs.

The example tests are run in CI so this ensures that our documentation examples don't go stale.

File: starlite/datastructures/headers.py (hunk 2 above)
Comment (seladb) [reply starlite/datastructures/headers.py:14]: I don't think `documentation_only` is needed in these "extra" headers. Do you see a use case where someone needs to add them only for documentation (OpenAPI) purposes?

Where do I mention it in the documentation?

**Review by peterschutt — APPROVED** [review]
Nice! Please see my review about how to include complete examples in the documentation.

File: starlite/datastructures/headers.py (hunk 2 above)
Comment (Goldziher) [reply starlite/datastructures/headers.py:14]: This is meant for the cases where users set headers dynamically. For example, when returning a response instance they create.

Comment (seladb) [reply starlite/datastructures/headers.py:14]: I'm not sure I understand... 

When users want to use `CacheControlHeader` they'll likely do something like this:

```python
@get(path="/handler", cache_control=CacheControlHeader(max_age=1000))
def my_handler(self) -> None:
    pass
```

Behind the scenes, [we create a `ResponseHeader` instance](https://github.com/seladb/starlite/blob/da0ef9943b2521564772aa7bc4d41ca8e8b90051/starlite/handlers/http.py#L433) with a default value of `documentation_only=False` and add it to the list of `resolved_response_headers`.

In this scenario, what's the use case for creating a `CacheControlHeader(..., documentation_only=True)`?

In other words, why would users want to create a `CacheControlHeader` header if they don't want to return it in the response?

File: starlite/controller.py [hunk 6]
```diff
@@ -64,6 +65,11 @@ class Controller:
         It receives the [Request][starlite.connection.Request] instance and any
         non-`None` return value is used for the response, bypassing the route handler.
     """
+    cache_control: Optional["CacheControlHeader"]
+    """
+         A [CacheControlHeader][starlite.datastructures.CacheControlHeader] header to add to route handlers
+         of this controller.
+    """
```
[the remark this answers is not in the document]
Comment (seladb) [reply starlite/controller.py:72]: Fixed here: https://github.com/starlite-api/starlite/pull/601/commits/7432876afd1d81e445a423ade5f7c9e326750491

File: starlite/config/app.py [hunk 7]
```diff
@@ -103,6 +103,11 @@ class Config(BaseConfig):
     """
     Configures caching behavior of the application.
     """
+    cache_control: Optional[CacheControlHeader]
+    """
+    A `cache-control` header of type [CacheControlHeader][starlite.datastructures.CacheControlHeader] to add to
+    route handlers of this app.
+    """
```
[the remark this answers is not in the document]
Comment (seladb) [reply starlite/config/app.py:110]: Fixed here: https://github.com/starlite-api/starlite/pull/601/commits/7432876afd1d81e445a423ade5f7c9e326750491

File: starlite/datastructures/headers.py (hunk 2 above)
Comment (provinzkraut) [reply starlite/datastructures/headers.py:14]: They might want to return it in the response, but the view could be dependent on the response itself, so they couldn’t be set at the time when they declare the handler.

In this case, having `documentation_only=True` means they can set that when declaring the handler, but set the "real header" dynamically within the function body.

> Where do I mention it in the documentation?

Disregard that please. You didn’t actually mention it in the docs, I just mixed up some lines there while reading.

Comment (seladb) [reply starlite/datastructures/headers.py:14]: Sure, added here: https://github.com/starlite-api/starlite/pull/601/commits/2089e9002542eee8d171517424d80bd77999a1fe

File: tests/response/test_response_headers.py (hunk 3 above)
Comment (seladb) [reply tests/response/test_response_headers.py:77]: Sure, the test was added here: https://github.com/starlite-api/starlite/pull/601/commits/56a8814ae9e7b1a30bc34b00d4236cbc82734fad

File: tests/datastructures/test_headers.py (hunk 4 above)
Comment (seladb) [reply tests/datastructures/test_headers.py:14]: I'm not sure there's a reason to test all possible fields because there's no per-field logic. The logic is per field type - either "boolean" fields (like `public`, `no-store`, etc.) or "integer" fields (like `max-age`). This test tests both field types

Comment (provinzkraut) [reply tests/datastructures/test_headers.py:14]: It would serve more to codify the field names, so changing their names / types would fail tests if not done in a backwards compatible way.

**Review by provinzkraut — COMMENTED** [review]
Looks good! Nothing more to add except the remaining open comment, but we can go either way about this.

File: docs/usage/5-responses/4-response-headers.md (hunk 5 above)
Comment (seladb) [reply docs/usage/5-responses/4-response-headers.md]: Thanks! Fixed here: https://github.com/starlite-api/starlite/pull/601/commits/7dc701b993ca6f097558cf2df6f32584ae6128da

**Review by provinzkraut — APPROVED** [review]

File: tests/datastructures/test_headers.py (hunk 4 above)
Comment (seladb) [reply tests/datastructures/test_headers.py:14]: Sure, fixed here: https://github.com/starlite-api/starlite/pull/601/commits/0f7dd0639e3401d984f94fc942036d191ce9aae7
